### PR TITLE
Add default cases to switch statements

### PR DESCRIPTION
**Pull request by xtofl**
The simplest client file, when compiling with `-Wall -Wextra`, results in a ton of unneeded warnings.

The test script I used:

```
#!/bin/bash

# build the package
(
  mkdir build
  cd build
  cmake ..
  make package
  tar -xzf dbus-cxx-0.1.1-Linux.tar.gz
)

# use the package
cat <<EOF > compile.cpp
#include <dbus-cxx.h>
#include <dbus-cxx/signal_base.h>
#include <dbus-cxx/dbus_signal.h>
int main() {}
EOF

export PKG_CONFIG_PATH=build/dbus-cxx-0.1.1-Linux/lib/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig

pkgconf -cflags dbus-cxx-1.0 dbus-1
export CXXFLAGS=$(pkgconf -cflags dbus-cxx-1.0 dbus-1) 

g++ -Wall -Wextra -Werror $CXXFLAGS -I build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11 compile.cpp
```

```
g++ -Wall $CXXFLAGS -I dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11 ../compile.cpp 
/home/krpi/dbus-cxx/dbus-cxx/methodbase.cpp:63:57: error: default argument given for parameter 1 of ‘std::__cxx11::string DBus::MethodBase::introspect(int) const’ [-fpermissive]
   std::string MethodBase::introspect(int space_depth=0) const
                                                         ^~~~~
In file included from /home/krpi/dbus-cxx/dbus-cxx/methodbase.cpp:20:0:
/home/krpi/dbus-cxx/dbus-cxx/methodbase.h:68:27: note: previous specification in ‘virtual std::__cxx11::string DBus::MethodBase::introspect(int) const’ here
       virtual std::string introspect(int space_depth=0) const;
                           ^~~~~~~~~~
CMakeFiles/dbus-cxx.dir/build.make:350: recipe for target 'CMakeFiles/dbus-cxx.dir/dbus-cxx/methodbase.cpp.o' failed
make[2]: *** [CMakeFiles/dbus-cxx.dir/dbus-cxx/methodbase.cpp.o] Error 1
CMakeFiles/Makefile2:109: recipe for target 'CMakeFiles/dbus-cxx.dir/all' failed
make[1]: *** [CMakeFiles/dbus-cxx.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
-I/usr/local/include/dbus-cxx-0.11 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/sigc++-2.0 -I/usr/lib/x86_64-linux-gnu/sigc++-2.0/include  
In file included from build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/interface.h:30:0,
                 from build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/object.h:705,
                 from build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/connection.h:33,
                 from build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx.h:29,
                 from compile.cpp:1:
build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/methodbase.h: In member function ‘virtual std::__cxx11::string DBus::MethodBase::introspect(int) const’:
build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/methodbase.h:68:54: error: unused parameter ‘space_depth’ [-Werror=unused-parameter]
       virtual std::string introspect(int space_depth=0) const { return std::string(); }
                                                      ^
build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/methodbase.h: In member function ‘virtual std::__cxx11::string DBus::MethodBase::arg_name(size_t)’:
build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/methodbase.h:70:43: error: unused parameter ‘i’ [-Werror=unused-parameter]
       virtual std::string arg_name(size_t i) { return std::string(); }
                                           ^
build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/methodbase.h: In member function ‘virtual void DBus::MethodBase::set_arg_name(size_t, const string&)’:
build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/methodbase.h:72:40: error: unused parameter ‘i’ [-Werror=unused-parameter]
       virtual void set_arg_name(size_t i, const std::string& name) { }
                                        ^
build/dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/methodbase.h:72:62: error: unused parameter ‘name’ [-Werror=unused-parameter]
       virtual void set_arg_name(size_t i, const std::string& name) { }


In file included from dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/messageiterator.h:20:0,
                 from dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/message.h:26,
                 from dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/callmessage.h:19,
                 from dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx.h:28,
                 from ../compile.cpp:1:
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h: In function ‘std::__cxx11::string DBus::include_file_for_type(DBus::Type)’:
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_INVALID’ not handled in switch [-Wswitch]
     switch ( t )
            ^
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_BOOLEAN’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_DOUBLE’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_OBJECT_PATH’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_SIGNATURE’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_VARIANT’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_STRUCT’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:172:12: warning: enumeration value ‘TYPE_UNIX_FD’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h: In function ‘bool DBus::type_is_templated(DBus::Type)’:
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_INVALID’ not handled in switch [-Wswitch]
     switch( t )
           ^
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_BYTE’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_BOOLEAN’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_INT16’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_UINT16’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_INT32’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_UINT32’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_INT64’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_UINT64’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_DOUBLE’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_STRING’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_OBJECT_PATH’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_SIGNATURE’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_VARIANT’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_STRUCT’ not handled in switch [-Wswitch]
dbus-cxx-0.1.1-Linux/include/dbus-cxx-0.11/dbus-cxx/types.h:193:11: warning: enumeration value ‘TYPE_UNIX_FD’ not handled in switch [-Wswitch]
```

### CODE REVIEW

**Comment by xtofl** [comment]
I forgot to fix the _other_ warnings we got in our build: unused parameter.  And I screwed up the tabs/spaces.  These issues have been appended now.

If need be, I can easily rebase this into two simple commits.

**Comment by rm5248** [comment]
I can squash the commits down when I merge, that's not a big issue.  I'm a little curious as to the moving of some of the implementation out of the header file, I don't think it breaks anything but I'm not sure - I'm worried about ABI compatibility.

**Comment by rm5248** [comment]
If this is also causing a problem immediately with your code, check this [SO question](https://stackoverflow.com/q/15053776/624483) to suppress the warnings from just the library.

**Comment by xtofl** [comment]
Regarding the move of the implementation into the source file, I tried to carefully think it through and put my reasoning in the respective commit message.

TL; DR; quite sure it's safe.

But please, don't trust it :).

## Regarding ABI compatibility

The pre-merge and post-merge signal_base.o both contain a definition of all virtual functions with a definition in the header; since the functions are all `virtual`.

Pre-merge client code also contains a definition.  It _should_ be the same, so if the client is not rebuilt, but uses a post-merge version of the library, it _should_ be allright.

https://stackoverflow.com/questions/603390/inline-member-functions-in-c

Note: I wonder if you really want to support clients to not rebuild after importing a new version of the library.  This constraint makes it _very_ hard to evolve the API.

**Comment by xtofl** [comment]
While using the library further, I found some other warnings that could be improved:

```
/usr/local/include/dbus-cxx-0.11/dbus-cxx/dbus_signal.h: In member function 'std::__cxx11::string DBus::signal<T_return, T_arg1, T_arg2, sigc::nil, sigc::nil, sigc::nil, sigc::nil, sigc::nil>::introspect(int) const [with T_return = void; T_arg1 = std::__cxx11::basic_string<char>; T_arg2 = int]':
/usr/local/include/dbus-cxx-0.11/dbus-cxx/dbus_signal.h:862:95: error: 'arg2' may be used uninitialized in this function [-Werror=maybe-uninitialized]
     sout << spaces << "  <arg name=\"" << m_arg_names[2-1] << "\" type=\"" << signature(arg2) << "\"/>\n";
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~

```

and

```
In file included from /usr/local/include/dbus-cxx-0.11/dbus-cxx/interface.h:488:0,
                 from /usr/local/include/dbus-cxx-0.11/dbus-cxx/object.h:705,
                 from /usr/local/include/dbus-cxx-0.11/dbus-cxx/connection.h:33,
                 from /usr/local/include/dbus-cxx-0.11/dbus-cxx.h:29,
...
/usr/local/include/dbus-cxx-0.11/dbus-cxx/method.h: In member function 'std::__cxx11::string DBus::Method<T_return, T_arg1>::introspect(int) const [with T_return = int; T_arg1 = std::__cxx11::basic_string<char>]':
/usr/local/include/dbus-cxx-0.11/dbus-cxx/method.h:231:12: error: 'type' may be used uninitialized in this function [-Werror=maybe-uninitialized]
       sout << spaces << "  <arg name=\"" << m_arg_names[0]
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            << "\" type=\"" << signature(type)
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            << "\" direction=\"out\"/>\n";
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Due to the use of the `signature` function.  Since the argument is only used for overload selection, this function _could_ become a function template.

**Comment by rm5248** [comment]
> Note: I wonder if you really want to support clients to not rebuild after importing a new version of the library. This constraint makes it _very_ hard to evolve the API.

I want to keep the ABI compatible for major releases; if you're curious you can look at the current status of the [dbus-cxx-2](https://github.com/dbus-cxx/dbus-cxx/tree/dbus-cxx-2) branch which breaks API and ABI compatibility to a certain extent.  So if there's a big API breaking change, it should go there.

> Due to the use of the `signature` function. Since the argument is only used for overload selection, this function _could_ become a function template.

It was in earlier versions, but this is problematic with the (current) version due to how the signatures need to be generated, and can get especially annoying with custom types due to the templates.  Since it needs to be recursive, the template can't handle this properly(as far as I am aware).  The issue comes up with complex data types, like `vector<vector<int>>`, which needs to be translated to a dbus type of `aai`.  If there's a better way, that would be useful for v2.
